sync_labels.py

```diff
@@ -0,0 +1,138 @@
+#!/usr/bin/env python3
+import sys
+import os
+import argparse
+from github import Github, GithubException
+
+"""
+Sync Github Labels
+
+Synchronize labels idempotently between two Github repositories.
+"""
+
+
+def sync_labels(src_repo, dest_repo, dry_run, delete):
+    """
+    Main subroutine: sync labels between source/destination Github repos
+
+    :param src_repo: full repo name for source of labels
+    :param dest_repo: full repo name for destination of label actions
+    :param dry_run: do a dry run, do not perform the actions
+    :param delete: delete labels that are in destination but not in source
+    """
+
+    # Get PyGithub objects representing repos
+    g = Github(get_access_token())
+    s = g.get_repo(src_repo)
+    d = g.get_repo(dest_repo)
+
+    # Create list of labels to add/delete
+    src_labels = list(s.get_labels())
+    src_label_names = set([j.name for j in src_labels])
+    dest_labels = list(d.get_labels())
+    dest_label_names = set([j.name for j in dest_labels])
+    dest_add = src_label_names - dest_label_names
+    dest_del = dest_label_names - src_label_names
+
+    # Show summary of what actions will be taken, get confirmation
+    print(f"The following labels will be created in {dest_repo}:")
+    print_list(list(dest_add), " [+] ")
+
+    if delete:
+        print(f"The following labels will be deleted in {dest_repo}:")
+        print_list(list(dest_del), " [-] ")
+
+    # Add new labels
+    add_ok = []
+    add_fail = []
+    for name in dest_add:
+        if dry_run:
+            add_ok.append(name)
+            continue
+        label = s.get_label(name)
+        color = label.color
+        description = label.description
+        try:
+            d.create_label(name, color, description)
+        except GithubException:
+            add_fail.append(name)
+        else:
+            add_ok.append(name)
+
+    # Delete labels
+    del_ok = []
+    del_fail = []
+    if delete:
+        for name in dest_del:
+            if dry_run:
+                del_ok.append(name)
+                continue
+            label = d.get_label(name)
+            try:
+                label.delete()
+            except GithubException:
+                del_fail.append(name)
```

Comment:
You do need to log the exception somehow. It is never a good idea to suppress a runtime error. If this exception occurs, the user needs to address the problem somehow. Hiding the cause makes that difficult.

Reply:
If you somehow expect an exception, then you need to be more specific about catching and suppressing exactly that exception (catching a more concrete subtype or interrogating the error code).

Reply:
I see your point about catching more specific exceptions. If the script fails part of the way through its work, one approach would be to have the script abort immediately without finishing the work (i.e., not handling the generic GithubException). Alternatively, the script could handle the generic exception, log the exception message, and keep going.

I opted for the second approach in my follow-up to this code review. But if the script is idempotent, I can see why the first approach would be preferred. Please let me know what you think.